scripts/redcap-extract.py

```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env python3
+import requests
+import io
+import csv
+from ftplib import FTP
+from datetime import datetime
+
+def extract_to_bytes(extract, fields=None):
+    """ Converts the REDCAP API response data to a io bytes handle
+    Paramters
+    _________
+    extract: list of dicts
+    fields: list of str
+        Field names to include in CSV.
+
+    Returns
+    _______
+    fb : io.BytesIO
+
+    """
+    return iostr_to_iobytes(write_csv(io.StringIO(''), extract, fields))
+
+def write_csv(f, extract, fields=None):
+    """ Writers extract to file handle
+    Parameters
+    __________
+    f : file handle (string mode)
+    extract: list of dicts
+    fields: list of str
+        Field names to include in CSV.
+
+    Returns
+    _______
+    f : file handle (string mode)
+    """
+    keys = fields if fields is not None else extract[0].keys()
+    dict_writer = csv.DictWriter(f, keys, extrasaction='ignore')
+    dict_writer.writeheader()
+    dict_writer.writerows(extract)
+    return f
+
+def iostr_to_iobytes(fs):
+    """ Converts io.StringIO object to io.BytesIO
+    Paramters
+    _________
+    fs : io.StringIO
+
+    Returns
+    _______
+    fb : io.BytesIO
+    """
+    fs.flush()
+    fs.seek(0)
+    fb = io.BytesIO(fs.read().encode())
+    return fb
+
+def bytes_to_ftp(f, project, name, ip, username='', password=''):
+    """ Saves file to the REDCAP extracts folder on RADAR-FTP.
+    Parameters
+    __________
+    f : file handle (bytes)
+    project : string
+        Name of the project folder to save under
+    name : string
+        Name to give the file
+    """
+    ftp = FTP(ip)
+    if (username and password):
+        ftp.login(username, password)
+    ftp.storbinary('STOR /RADAR-CNS/REDCAP/{}/{}'.format(project, name), f)
+    ftp.close()
+    return
```

Comment:
Is it possible to add support for SFTP (key-based too)?
Since it is now the default way we provide access to the NAS.

Reply:
The ftp std lib in python doesn't support SFTP. I have another script, which I actually use instead of this, which uses subprocess to call ssh. This script is more portable though (doesn't depend on openSSH, uses only stdlib if I remove the requests dependency)